### DISCUSSION TITLE
What is the Video Streaming Protocol used?

### DISCUSSION THREAD

**Discussion by sumanthrf** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#1029] [Q&A] [answered]
What are the protocols this SDK supports? We know this supports HLS. But does it also support LL-HLS , MPEG-DASH , WebRTC or SRT?

**Comment by hassanctech** [comment, marked as answer]
This SDK supports the ingestion of video data into the Kinesis Video Streams backend service, hence it is agnostic to the streaming protocol used on the consumption side.  

On the consumption side there are 3 ways to get your video data:
1. HLS [https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/hls-playback.html]
2. MPEG-DASH [https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/dash-playback.html]
3. GetMedia API [https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_dataplane_GetMedia.html]

LL-HLS is not currently supported.

For WebRTC we have a separate class of SDKs:
1. C SDK [https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c]
2. JS SDK [https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js]
3. Android SDK [https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-android]
4. iOS SDK [https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-ios]

This is a feature which is currently in public preview which supports ingestion of video stream data via WebRTC, to see how this works you can consult this sample:

https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/tree/Release-WebRTC-Streams-Ingestion-Public-Preview#running-the-samples, look at `kvsWebRTCClientBasicStorageSession`.  As it stands this uses WebRTC to ingest the video into streams, but today this path does not support a traditional WebRTC viewer, it only supports the SDK sending video and the backend persisting it a Kinesis Video Stream.  From there to consume the data you must use one of the 3 consumption paths outlined at the top.

**Comment by reimida** [comment]
which protocol is used for real time ingestion (the producer side), RTP?

**Reply by sirknightj** [reply]
Producer SDK's send media to KVS using HTTPS (specifically, https://en.wikipedia.org/wiki/Chunked_transfer_encoding) of [MKV data](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-data.html) via [PutMedia API](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_dataplane_PutMedia.html).

If you have a camera using RTSP, for example, you can extract the h.264/5 frames out, and pass it to the SDK via putKinesisVideoFrame/putFrame, which will handle packaging it into MKV and uploading it to KVS.

If you want to send RTP packets to KVS for ingestion and storage, then [WebRTC ingestion](https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/webrtc-ingestion.html) is what is currently available.

**Reply by reimida** [reply]
thank you brother. at the moment, i am using the edge connector for kvs and it is surplisingly fast: it uses put media api which probably is based on putframe? How is it possible with a protocol like http to have a delay tipically <1s? is it worth to switch to webrtc, which is faster?

one final question and i'm done: since the consumption side is agnostic of the protocol on the producer side, the whole stuff would still work even switching to webrtc, without any change? because at the moment i am using a library based on get media api: https://github.com/aws-samples/amazon-kinesis-video-streams-consumer-library-for-python/tree/main
